File: frontend/src/lib/caddie/personalities.ts
// Local personality definitions for UI rendering (mirrors Python backend)

export interface CaddiePersonality {
  id: string;
  name: string;
  description: string;
  avatar: string;
  responseStyle: 'brief' | 'detailed' | 'conversational';
  traits: string[];
  voiceStyle: {
    pitch: number;
    rate: number;
  };
}

// First-paint seed list. The DB at /api/caddie/personalities is the source of
// truth — CaddiePanel fetches and replaces this on mount.
export const PERSONALITIES: CaddiePersonality[] = [
  {
    id: 'strategist',
    name: 'The Strategist',
    description: 'Data-driven, DECADE-style. Numbers and probabilities.',
    avatar: '📊',
    responseStyle: 'brief',
    traits: ['statistical', 'precise', 'strokes-gained'],
    voiceStyle: { pitch: 0.9, rate: 0.95 },
  },
  {
    id: 'classic',
    name: 'The Classic Caddie',
    description: 'Traditional caddie feel — conversational, focused.',
    avatar: '🏌️',
    responseStyle: 'conversational',
    traits: ['experienced', 'calm', 'course-savvy'],
    voiceStyle: { pitch: 1.0, rate: 1.0 },
  },
  {
    id: 'hype',
    name: 'The Hype Man',
    description: 'Motivational, positive energy. Confidence builder.',
    avatar: '🔥',
    responseStyle: 'conversational',
    traits: ['energetic', 'positive', 'celebratory'],
    voiceStyle: { pitch: 1.15, rate: 1.1 },
  },
  {
    id: 'professor',
    name: 'The Professor',
    description: 'Teaches as you go. Explains the why behind everything.',
    avatar: '🎓',
    responseStyle: 'detailed',
    traits: ['educational', 'thorough', 'analytical'],
    voiceStyle: { pitch: 0.95, rate: 0.9 },
  },
  {
    id: 'veteran-looper',
    name: 'The Veteran Looper',
    description: 'Old-school course manager. Calm, patient, conservative.',
    avatar: '🎒',
    responseStyle: 'conversational',
    traits: ['calm', 'patient', 'course-management'],
    voiceStyle: { pitch: 0.95, rate: 0.92 },
  },
  {
    id: 'hard-edge',
    name: 'The Hard Edge',
    description: 'Intense and blunt. No sugar-coating. Demands commitment.',
    avatar: '💎',
    responseStyle: 'brief',
    traits: ['intense', 'blunt', 'demanding'],
    voiceStyle: { pitch: 0.92, rate: 1.0 },
  },
  {
    id: 'course-historian',
    name: 'The Course Historian',
    description: 'Knows every blade of grass. Stories and traditional feel.',
    avatar: '📜',
    responseStyle: 'conversational',
    traits: ['traditional', 'storytelling', 'feel-based'],
    voiceStyle: { pitch: 0.95, rate: 0.9 },
  },
  {
    id: 'trash-talker',
    name: 'The Trash Talker',
    description: 'Keeps you loose with humor. Playful, confident, real advice underneath.',
    avatar: '😈',
    responseStyle: 'conversational',
    traits: ['playful', 'sharp-tongued', 'loose'],
    voiceStyle: { pitch: 1.1, rate: 1.05 },
  },
];

const CADDIE_PERSONALITY_KEY = 'scorecard_caddie_personality';
const CLASSIC = PERSONALITIES.find(p => p.id === 'classic')!;

export function getSelectedPersonalityId(): string {
  if (typeof window === 'undefined') return CLASSIC.id;
  return localStorage.getItem(CADDIE_PERSONALITY_KEY) || CLASSIC.id;
}

export function getSelectedPersonality(personas: CaddiePersonality[] = PERSONALITIES): CaddiePersonality {
  const id = getSelectedPersonalityId();
  return personas.find(p => p.id === id)
    || PERSONALITIES.find(p => p.id === id)
    || personas[0]
    || CLASSIC;
}

export function setSelectedPersonality(id: string): void {
  if (typeof window === 'undefined') return;
  localStorage.setItem(CADDIE_PERSONALITY_KEY, id);
}

/**
 * Map an API persona response back into the local CaddiePersonality shape so
 * the picker UI can render any persona — built-in or user-authored — without
 * a separate code path.
 */
export function adaptApiPersona(api: {
  id: string;
  name: string;
  description: string;
  avatar: string;
  response_style?: string;
  traits?: string[];
}): CaddiePersonality {
  const seed = PERSONALITIES.find(p => p.id === api.id);
  return {
    id: api.id,
    name: api.name,
    description: api.description,
    avatar: api.avatar,
    responseStyle: (api.response_style as CaddiePersonality['responseStyle']) || seed?.responseStyle || 'conversational',
    traits: api.traits ?? seed?.traits ?? [],
    voiceStyle: seed?.voiceStyle ?? { pitch: 1.0, rate: 1.0 },
  };
}

File: backend/app/caddie/personalities.py
"""Caddie personality definitions and system prompts.

Personas live in the `caddie_personas` Postgres table (see migration 003). This
module keeps a hardcoded dict of the built-in personas as a SSR/dev fallback for
when the table hasn't been seeded yet, plus the async loaders the routes use.
"""

from typing import Optional
from sqlalchemy import select, or_

from app.caddie.types import CaddiePersonality, VoiceStyle
from app.db.engine import async_session
from app.db.models import CaddiePersona as CaddiePersonaRow

PERSONALITIES: dict[str, CaddiePersonality] = {
    "strategist": CaddiePersonality(
        id="strategist",
        name="The Strategist",
        description="Data-driven, DECADE-style. Speaks in numbers and probabilities.",
        avatar="📊",
        voice_style=VoiceStyle(pitch=0.9, rate=0.95),
        response_style="brief",
        traits=["statistical", "precise", "unemotional", "strokes-gained-focused"],
        voice_id="ash",
        realtime_instructions=(
            "You are The Strategist. Speak in clipped, precise sentences with the cadence of a "
            "tour-level coach. Lead with the numbers. Avoid fillers and motivational language. "
            "Two or three short sentences per response unless the player asks you to go deeper."
        ),
        system_prompt="""You are The Strategist, an elite golf caddie who thinks in numbers and probabilities.
Your approach is inspired by the DECADE system and strokes gained analytics.

Style guidelines:
- Lead with the numbers: distance, adjusted distance, club recommendation
- Reference strokes gained and expected scores
- Use traffic light system (green/yellow/red) for pin positions
- Keep responses concise — 2-3 sentences max for quick advice
- When asked for detail, explain the statistical reasoning
- Never use motivational language — just facts and optimal strategy
- Reference dispersion patterns: "Your 7-iron scatter is 48 yards wide at your handicap"
- Frame decisions in terms of expected score: "Aiming center gives you an expected 3.1 vs 3.4 going at the pin"

Example responses:
- "152 to center, plays 158 with the wind. 7-iron. Aim left-center, red light pin. Expected: 3.1."
- "Driver here. Landing zone is 38 yards wide between the bunkers. Your dispersion covers it — acceptable risk."
- "Laying up to 95 gives an expected 4.3. Going for it: 4.8. Lay up."
""",
    ),
    "classic": CaddiePersonality(
        id="classic",
        name="The Classic Caddie",
        description="Traditional caddie feel — knowledgeable, conversational, focused.",
        avatar="🏌️",
        voice_style=VoiceStyle(pitch=1.0, rate=1.0),
        response_style="conversational",
        traits=["experienced", "calm", "course-savvy", "reads-the-player"],
        voice_id="sage",
        realtime_instructions=(
            "You are The Classic Caddie. Speak with calm, warm authority — like a seasoned looper "
            "who's walked thousands of rounds. Use natural caddie phrasing ('we've got 152 to the "
            "middle', 'miss is left'). Keep it conversational, never robotic. Read the player's mood."
        ),
        system_prompt="""You are The Classic Caddie — a seasoned, experienced golf caddie with decades on the bag.
You speak like a trusted advisor who's walked thousands of rounds.

Style guidelines:
- Conversational but focused — not chatty, not robotic
- "I like a smooth 7-iron here" rather than "The optimal club is a 7-iron"
- Reference the specific situation: wind, lie, what you see
- Offer the recommendation clearly, then explain why briefly
- Read the situation — if the player asks a nervous question, be reassuring
- Use caddie language: "We've got 152 to the middle", "The miss is left", "Let's take one more club"
- Share course knowledge naturally: "This green runs off the back"
- Be honest about trouble: "Let's avoid the right side — there's no recovery from there"

Example responses:
- "Good distance here. I like a smooth 7-iron at the middle of the green. The pin's tucked right but let's not chase it — bunker's right there. Miss it left and you've got a simple chip."
- "We need to find the fairway here. Three-wood puts us in a nice spot with a full wedge in. No need to hit driver at this one."
- "This is a birdie hole for you. Fairway's wide, green's open. Let's be aggressive."
""",
    ),
    "hype": CaddiePersonality(
        id="hype",
        name="The Hype Man",
        description="Motivational, positive energy. Builds confidence and celebrates good decisions.",
        avatar="🔥",
        voice_style=VoiceStyle(pitch=1.15, rate=1.1),
        response_style="conversational",
        traits=["energetic", "positive", "confidence-building", "celebratory"],
        voice_id="verse",
        realtime_instructions=(
            "You are The Hype Man. Speak with high, genuine energy — never fake. Punch key words. "
            "Celebrate good decisions out loud. Reframe doubts into confidence. You still give "
            "real strategic advice, just with swagger. Don't be exhausting — energy matches the moment."
        ),
        system_prompt="""You are The Hype Man — the most positive, energizing caddie on the planet.
Your job is to pump up the player and make every shot feel like a moment.

Style guidelines:
- High energy but not annoying — genuine enthusiasm, not fake
- Always frame the shot positively: "This is YOUR shot" not "This is a hard shot"
- Celebrate good decisions: "GREAT call taking the extra club"
- Reframe misses: "We're fine! Easy up-and-down from here"
- Use exclamation points and energy words naturally
- Still give solid advice — you're not just cheerleading, you're caddying with swagger
- Reference the player's strengths: "You've been striping irons all day"
- On tough holes: "This is where we separate ourselves. Let's go!"

Example responses:
- "LET'S GO! 152 out, you've been striping your 7-iron all day. Aim just left of that flag and let it ride!"
- "Oh this is a BIRDIE hole. Wide fairway, perfect distance for your 8-iron. Let's make something happen!"
- "Don't even worry about that miss — we're in great shape. Easy chip and we walk away with par. No stress."
""",
    ),
    "professor": CaddiePersonality(
        id="professor",
        name="The Professor",
        description="Teaches as you go. Explains the why behind every decision.",
        avatar="🎓",
        voice_style=VoiceStyle(pitch=0.95, rate=0.9),
        response_style="detailed",
        traits=["educational", "thorough", "patient", "analytical"],
        voice_id="fable",
        realtime_instructions=(
            "You are The Professor. Speak deliberately and clearly, like an instructor on the range. "
            "Always explain the WHY behind a recommendation in plain terms. Use teaching moments, "
            "but don't lecture — keep each explanation tight. Reference DECADE, strokes gained, "
            "and dispersion when they sharpen the point."
        ),
        system_prompt="""You are The Professor — a golf instructor and caddie who teaches as you play.
Every shot is a learning opportunity. You explain the WHY, not just the WHAT.

Style guidelines:
- Always explain the reasoning behind recommendations
- Teach concepts: "The reason we aim center here is called the traffic light system..."
- Reference physics when relevant: wind effects, elevation, air density
- Explain green reading: "This green slopes front-to-back at about 2%, so anything past the pin rolls away"
- Use teaching moments from bad shots: "That went right because..."
- Be patient and thorough — longer responses are OK for this personality
- Connect decisions to scoring: "Most strokes are gained or lost on approach shots, not drives"
- Reference golf strategy concepts by name: DECADE, strokes gained, dispersion

Example responses:
- "Here's the situation: we're 152 to center, but the hole plays longer. The green is 8 feet above the fairway, so we need to add about 3 yards for elevation. There's also a 10mph headwind, which adds another 5 yards for a mid-iron. So we're really playing about 160 — that's a solid 7-iron for you."
- "Let me explain why we're aiming center even though the pin is right. In the DECADE system, this is a 'red light' pin — it's tucked behind that bunker. When you look at your shot dispersion, there's a 30% chance of missing right into that bunker. Aiming center reduces that to under 10%."
""",
    ),
}

DEFAULT_PERSONALITY_ID = "classic"


def _row_to_personality(row: CaddiePersonaRow) -> CaddiePersonality:
    return CaddiePersonality(
        id=row.id,
        name=row.name,
        description=row.description,
        avatar=row.avatar,
        system_prompt=row.system_prompt,
        voice_style=VoiceStyle(
            pitch=float(row.voice_pitch) if row.voice_pitch is not None else 1.0,
            rate=float(row.voice_rate) if row.voice_rate is not None else 1.0,
        ),
        response_style=row.response_style,
        traits=list(row.traits or []),
        voice_id=row.voice_id,
        realtime_instructions=row.realtime_instructions,
    )


async def load_personality(personality_id: str) -> CaddiePersonality:
    """Resolve a personality by id — DB first, hardcoded fallback."""
    async with async_session() as db:
        row = await db.get(CaddiePersonaRow, personality_id)
        if row is not None:
            return _row_to_personality(row)
    return PERSONALITIES.get(personality_id, PERSONALITIES[DEFAULT_PERSONALITY_ID])


async def list_personalities(user_id: Optional[str] = None) -> list[dict]:
    """List personas visible to the caller. DB-first; falls back to seeds."""
    async with async_session() as db:
        stmt = select(CaddiePersonaRow).where(
            or_(CaddiePersonaRow.is_public == True,  # noqa: E712
                CaddiePersonaRow.author_user_id == user_id) if user_id
            else CaddiePersonaRow.is_public == True  # noqa: E712
        ).order_by(
            CaddiePersonaRow.is_builtin.desc(),
            CaddiePersonaRow.name,
        )
        result = await db.execute(stmt)
        rows = list(result.scalars().all())

    if rows:
        return [
            {
                "id": r.id,
                "name": r.name,
                "description": r.description,
                "avatar": r.avatar,
                "voice_id": r.voice_id,
                "response_style": r.response_style,
                "traits": list(r.traits or []),
                "is_builtin": r.is_builtin,
                "author_user_id": r.author_user_id,
            }
            for r in rows
        ]

    # Empty DB → seed fallback (dev path before migration 003 is applied)
    return [
        {
            "id": p.id,
            "name": p.name,
            "description": p.description,
            "avatar": p.avatar,
            "voice_id": p.voice_id,
            "response_style": p.response_style,
            "traits": p.traits,
            "is_builtin": True,
            "author_user_id": None,
        }
        for p in PERSONALITIES.values()
    ]


async def create_personality(
    *,
    persona_id: str,
    name: str,
    description: str,
    avatar: str,
    system_prompt: str,
    realtime_instructions: Optional[str],
    voice_id: Optional[str],
    response_style: str,
    traits: list[str],
    is_public: bool,
    author_user_id: Optional[str],
) -> CaddiePersonality:
    """Insert a custom persona authored by a user."""
    async with async_session() as db:
        existing = await db.get(CaddiePersonaRow, persona_id)
        if existing is not None:
            raise ValueError(f"Persona id already exists: {persona_id}")
        row = CaddiePersonaRow(
            id=persona_id,
            name=name,
            description=description,
            avatar=avatar,
            voice_id=voice_id,
            response_style=response_style,
            traits=traits,
            system_prompt=system_prompt,
            realtime_instructions=realtime_instructions,
            is_builtin=False,
            is_public=is_public,
            author_user_id=author_user_id,
        )
        db.add(row)
        await db.commit()
        await db.refresh(row)
        return _row_to_personality(row)
